File: UnityProject/Assets/Scripts/Data/FDiceDataManager.cs
using FEnum;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public class FDiceData
{
    public readonly int id;
    public readonly DiceGrade grade;
    public readonly string name;
    public readonly string description;
    public readonly string iconPath;
    public readonly string notAcquiredIconPath;
    public readonly Color color;
    public readonly List<int> skillIDList = new List<int>();

    public FDiceData(FDataNode InNode)
    {
        id = InNode.GetIntAttr("id");
        grade = (DiceGrade)InNode.GetIntAttr("grade");
        name = InNode.GetStringAttr("name");
        description = InNode.GetStringAttr("description");
        iconPath = InNode.GetStringAttr("icon");
        notAcquiredIconPath = InNode.GetStringAttr("notAcquiredIcon");
        color = InNode.GetColorAttr("color");

        FDataNode skillListNode = InNode.FindChildNode("SkillList");
        if(skillListNode != null)
        {
            skillListNode.ForeachChildNodes("Skill", (in FDataNode InNode) => {
                skillIDList.Add(InNode.GetIntAttr("id"));
            });
        }
    }

    public delegate void ForeachSkillIDDelegate(int InID);
    public void ForeachSkillID(ForeachSkillIDDelegate InFunc)
    {
        foreach(int id in skillIDList)
        {
            InFunc(id);
        }
    }
}

public class FDiceLevelData
{
    public readonly int level;
    public readonly int diceCountCost;
    public readonly int goldCost;

    public FDiceLevelData(FDataNode InNode)
    {
        level = InNode.GetIntAttr("level");
        diceCountCost = InNode.GetIntAttr("diceCountCost");
        goldCost = InNode.GetIntAttr("goldCost");
    }
}

public class FDiceGradeData
{
    public readonly DiceGrade grade;
    public readonly string gradeName;
    public readonly string backgroundPath;
    public readonly int initialLevel;
    public readonly int maxLevel;
    public readonly int critical;

    private Dictionary<int, FDiceLevelData> levelDataMap = new Dictionary<int, FDiceLevelData>();

    public FDiceGradeData(FDataNode InNode)
    {
        grade = (DiceGrade)InNode.GetIntAttr("grade");
        gradeName = InNode.GetStringAttr("name");
        backgroundPath = InNode.GetStringAttr("invenSlotImage");
        initialLevel = InNode.GetIntAttr("initialLevel");
        critical = InNode.GetIntAttr("critical");

        InNode.ForeachChildNodes("Level", (in FDataNode InNode) => {
            FDiceLevelData levelData = new FDiceLevelData(InNode);
            levelDataMap.Add(levelData.level, levelData);
        });

        maxLevel = levelDataMap.Keys.Max();
    }

    public FDiceLevelData FindDiceLevelData(int InLevel)
    {
        if (levelDataMap.ContainsKey(InLevel))
            return levelDataMap[InLevel];

        return null;
    }
}

public class FDiceDataManager : FNonObjectSingleton<FDiceDataManager>
{
    Dictionary<int, FDiceData> diceDataMap = new Dictionary<int, FDiceData>();
    Dictionary<DiceGrade, FDiceGradeData> diceGradeDataMap = new Dictionary<DiceGrade, FDiceGradeData>();

    public void Initialize()
    {
        List<FDataNode> diceDataNodes = FDataCenter.Instance.GetDataNodesWithQuery("DiceList.Dice");
        foreach(FDataNode node in diceDataNodes)
        {
            FDiceData data = new FDiceData(node);
            diceDataMap.Add(data.id, data);
        }

        List<FDataNode> diceGradeDataNodes = FDataCenter.Instance.GetDataNodesWithQuery("DiceGradeList.DiceGrade");
        foreach (FDataNode node in diceGradeDataNodes)
        {
            FDiceGradeData gradeData = new FDiceGradeData(node);
            diceGradeDataMap.Add(gradeData.grade, gradeData);
        }
    }

    public delegate void ForeachDiceDataFunc(in FDiceData InDiceData);
    public void ForeachDiceData(in ForeachDiceDataFunc InFunc)
    {
        foreach (FDiceData data in diceDataMap.Values)
        {
            InFunc(data);
        }
    }

    public FDiceData FindDiceData(int InID)
    {
        if (diceDataMap.ContainsKey(InID))
            return diceDataMap[InID];

        return null;
    }

    public FDiceGradeData FindGradeDataByID(int InID)
    {
        FDiceData diceData = FindDiceData(InID);
        if (diceData != null)
            return FindGradeData(diceData.grade);

        return null;
    }

    public FDiceGradeData FindGradeData(DiceGrade InGrade)
    {
        if(diceGradeDataMap.ContainsKey(InGrade))
            return diceGradeDataMap[InGrade];

        return null;
    }

    public FDiceLevelData FindDiceLevelData(int InID, int InLevel)
    {
        FDiceGradeData diceGradeData = FindGradeDataByID(InID);
        if (diceGradeData == null)
            return null;

        return diceGradeData.FindDiceLevelData(InLevel);
    }
}

File: Server/Server/Server/FDiceDataManager.cpp
#include "stdafx.h"
#include "FDiceDataManager.h"

void FDiceDataManager::Initialize()
{
	for (int i = DICE_GRADE_NORMAL; i < DICE_GRADE_MAX; ++i)
	{
		m_DiceIDByGradeMap.insert(pair<DiceGrade, vector<int>>((DiceGrade)i, vector<int>()));
	}

	vector<const FDataNode*> diceNodes = FDataCenter::GetInstance()->FindNodes("DiceList.Dice");
	for (const FDataNode* diceNode : diceNodes)
	{
		FDiceData diceData;
		diceData.id = diceNode->GetIntAttr("id");
		diceData.grade = (DiceGrade)diceNode->GetIntAttr("grade");

		m_DiceMap.insert(pair<int, FDiceData>(diceData.id, diceData));

		auto iter = m_DiceIDByGradeMap.find(diceData.grade);
		if (iter != m_DiceIDByGradeMap.end())
			iter->second.push_back(diceData.id);
	}

	vector<const FDataNode*> gradeNodes = FDataCenter::GetInstance()->FindNodes("DiceGradeList.DiceGrade");
	for (const FDataNode* gradeNode : gradeNodes)
	{
		FDiceGradeData data;
		data.grade = (DiceGrade)gradeNode->GetIntAttr("grade");
		data.initLevel = gradeNode->GetIntAttr("initialLevel");
		data.critical = gradeNode->GetIntAttr("critical");

		m_DiceGradeMap.insert(pair<DiceGrade, FDiceGradeData>(data.grade, data));

		gradeNode->ForeachChildNodes("Level", [&](const FDataNode* InNode) {
			FDiceGradeLevelData gradeLevelData;
			gradeLevelData.level = InNode->GetIntAttr("level");
			gradeLevelData.diceCountCost = InNode->GetIntAttr("diceCountCost");
			gradeLevelData.goldCost = InNode->GetIntAttr("goldCost");

			int id = ConvertDiceGradeLevelToID(data.grade, gradeLevelData.level);
			m_DiceGradeLevelMap.insert(pair<int, FDiceGradeLevelData>(id, gradeLevelData));
			});
	}
}

void FDiceDataManager::Release()
{
}

const FDiceData* FDiceDataManager::FindDiceData(int InID) const
{
	const auto& iter = m_DiceMap.find(InID);
	if (iter != m_DiceMap.end())
		return &iter->second;

	return nullptr;
}

const FDiceGradeData* FDiceDataManager::FindDiceGradeData(DiceGrade InGrade) const
{
	const auto& iter = m_DiceGradeMap.find(InGrade);
	if (iter != m_DiceGradeMap.end())
		return &iter->second;

	return nullptr;
}

const FDiceGradeData* FDiceDataManager::FindDiceGradeDataByID(int InID) const
{	
	if (const FDiceData* diceData = FindDiceData(InID))
	{
		const auto& iter = m_DiceGradeMap.find(diceData->grade);
		if (iter != m_DiceGradeMap.end())
			return &iter->second;
	}

	return nullptr;
}

const FDiceGradeLevelData* FDiceDataManager::FindDiceGradeLevelData(DiceGrade InGrade, int InLevel) const
{
	auto iter = m_DiceGradeLevelMap.find(ConvertDiceGradeLevelToID(InGrade, InLevel));
	if (iter != m_DiceGradeLevelMap.end())
		return &iter->second;

	return nullptr;
}

const FDiceGradeLevelData* FDiceDataManager::FindDiceGradeLevelDataByID(int InID, int InLevel) const
{
	if (const FDiceGradeData* gradeData = FindDiceGradeDataByID(InID))
	{
		return FindDiceGradeLevelData(gradeData->grade, InLevel);
	}

	return nullptr;
}

int FDiceDataManager::GetDiceCountByGrade(DiceGrade InGrade) const
{
	auto iter = m_DiceIDByGradeMap.find(InGrade);
	if (iter != m_DiceIDByGradeMap.end())
		return iter->second.size();

	return 0;
}

int FDiceDataManager::GetDiceIDByGrandAndIndex(DiceGrade InGrade, int InIndex) const
{
	auto iter = m_DiceIDByGradeMap.find(InGrade);
	if (iter == m_DiceIDByGradeMap.end())
		return 0;

	if (InIndex < 0 || iter->second.size() <= InIndex)
		return 0;

	return iter->second[InIndex];
}

int FDiceDataManager::ConvertDiceGradeLevelToID(DiceGrade InGrade, int InLevel) const
{
	return (InGrade << 8) + InLevel;
}